android/clientlib/src/main/java/com/solana/mobilewalletadapter/clientlib/scenario/LocalAssociationScenario.java

```diff
@@ -109,85 +110,98 @@ public Intent createAssociationIntent() {
+                        future = closeDeferredFuture();
+                        mMobileWalletAdapterWebSocket.close();
+                    } else {
+                        // In the middle of a backoff - we can tear down immediately
+                        mState = State.CLOSED;
+                        destroyResourcesOnClose();
+                        future = closedImmediatelyFuture();
+                    }
+                    break;
+                case ESTABLISHING_SESSION:
+                    mState = State.CLOSING;
+                    notifySessionEstablishmentFailed("Scenario closed during session establishment");
                     future = closeDeferredFuture();
                     mMobileWalletAdapterWebSocket.close();
-                } else {
-                    // In the middle of a backoff - we can tear down immediately
-                    mState = State.CLOSED;
-                    mHandler.removeCallbacksAndMessages(null); // pre-closed callbacks no longer relevant
-                    destroyResources();
+                    break;
+                case STARTED:
+                    mState = State.CLOSING;
+                    future = closeDeferredFuture();
+                    mMobileWalletAdapterWebSocket.close();
+                    break;
+                case CLOSING:
+                    // Already closing - nothing to do here
+                    future = closeDeferredFuture();
+                    break;
+                case CLOSED:
+                    // No-op; scenario is either already complete, or in the process of being torn down
                     future = closedImmediatelyFuture();
-                }
-                break;
-            case ESTABLISHING_SESSION:
-                mState = State.CLOSING;
-                notifySessionEstablishmentFailed("Scenario closed during session establishment");
-                future = closeDeferredFuture();
-                mMobileWalletAdapterWebSocket.close();
-                break;
-            case STARTED:
-                mState = State.CLOSING;
-                future = closeDeferredFuture();
-                mMobileWalletAdapterWebSocket.close();
-                break;
-            case CLOSING:
-                // Already closing - nothing to do here
-                future = closeDeferredFuture();
-                break;
-            case CLOSED:
-                // No-op; scenario is either already complete, or in the process of being torn down
-                future = closedImmediatelyFuture();
-                break;
-            default:
-                throw new IllegalStateException("Error: attempt to close in an unknown state");
+                    break;
+                default:
+                    throw new IllegalStateException("Error: attempt to close in an unknown state");
+            }
         }
 
         return future;
     }
 
+    @NonNull
+    @GuardedBy("mLock")
```

Comment:
What does this achieve from an operational perspective?

Reply:
Also, in case you hadn't noticed, I'm not too familiar with advanced operations relating to the JVM's threading model. But that's because I haven't had to for the most part. 99% of the time in my Android app experience, relatively simple thread management is enough, as long as I'm keeping the Android lifecycle in mind. I just want to make sure we are enabling as simple operations as we can for our consumers.

Reply:
Final point: I'm not saying this is wrong. I just wanna make sure my simpleton brain can understand this stuff. If so, then we are probably on the right track 😁

Reply:
The `@GuardedBy` annotation is for static analysis tools - it informs the tool what lock should be held when this method is called. It gives hints to call flow analysis tools to catch calls to methods without holding a required lock.

This is purely an implementation detail of LocalAssociationScenario; with this PR, none of the threading details leak out through the public APIs.